Inframanager policy stitching and control plane changes

venkateshpnv opened the pull request:
The changes include policy add/update, delete, workload add/update, delete in api_handler, control plane and store.

dhruvpt commented:
Shouldn't all these routes be using P4TAP_1 interface. The P4TAP_0 would not be available for Infra-agent as it is used only for running ARP proxy.
Further, why is the ServiceCIDR route entry removed? Won't we need access to services from host?

nupurjai reviewed (commented):
I guess, the P4TAP_0 in comments for host routes is a mistake in just comments. P4TAP_1 is in host space.

nupurjai commented:
Without an IP on host interface, the routes are rejected by ip route. Can you please add the IP address to P4TAP_1 in host namespace.

ahalimx86 commented:
> Shouldn't all these routes be using P4TAP_1 interface. The P4TAP_0 would not be available for Infra-agent as it is used only for running ARP proxy.

I didn't have the  P4TAP_0 removed or renamed prior to testing this our, that's why it picked this interface. Once P4TAP_0 is taken as suggested by you and Nupur, host will get next available interface (i.e., P4TAP_1)

> Further, why is the ServiceCIDR route entry removed? Won't we need access to services from host?

Nupur explicitly asked to remove ServiceCIDR. @nupurjai Can you please comment?

ahalimx86 commented:
> Without an IP on host interface, the routes are rejected by ip route. Can you please add the IP address to P4TAP_1 in host namespace.

The host interface gets assigned an IP (200.1.1.2).

```
39: P4TAP_0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1280 qdisc fq_codel state UNKNOWN group default qlen 1000                                                                                
    link/ether f6:56:48:f3:1a:51 brd ff:ff:ff:ff:ff:ff                                                                                                                                      
    inet 200.1.1.2/32 scope global P4TAP_0                                                                                                                                                  
       valid_lft forever preferred_lft forever 
```

venkateshpnv committed:
Inframanager policy stitching and control plane changes

Signed-off-by: Neelakanta Venkatesh Petla <neelakanta.venkatesh.petla@intel.com>

venkateshpnv committed:
Merge branch 'main' of https://github.com/venkateshpnv/k8s-infra-offload-new into mgr_policy

venkateshpnv committed:
Fix format errors

Signed-off-by: Neelakanta Venkatesh Petla <neelakanta.venkatesh.petla@intel.com>

venkateshpnv committed:
policy code fixes

Signed-off-by: Neelakanta Venkatesh Petla <neelakanta.venkatesh.petla@intel.com>

venkateshpnv committed:
Use macros for proto instead of numbers

Signed-off-by: Neelakanta Venkatesh Petla <neelakanta.venkatesh.petla@intel.com>

venkateshpnv committed:
Merge branch 'main' of https://github.com/venkateshpnv/k8s-infra-offload-new into mgr_policy

Signed-off-by: Neelakanta Venkatesh Petla <neelakanta.venkatesh.petla@intel.com>

venkateshpnv committed:
Fix build errors after the merge

Signed-off-by: Neelakanta Venkatesh Petla <neelakanta.venkatesh.petla@intel.com>

venkateshpnv committed:
Temporarily use dpdk cplane policy code for es2k

Signed-off-by: Neelakanta Venkatesh Petla <neelakanta.venkatesh.petla@intel.com>

venkateshpnv committed:
Fix policy test errors

Signed-off-by: Neelakanta Venkatesh Petla <neelakanta.venkatesh.petla@intel.com>

dhruvpt commented:
Please do not merge this change before 23.07 release. At present, only bug fixes should go in. Because, any such large scale changes would require us to re-run scans that we have already completed.

venkateshpnv committed:
Fixed failures in the policy code

Signed-off-by: Neelakanta Venkatesh Petla <neelakanta.venkatesh.petla@intel.com>

venkateshpnv committed:
Merge branch 'main' of https://github.com/venkateshpnv/k8s-infra-offload-new into mgr_policy

venkateshpnv committed:
Added provision for non tcp/udp traffic in policy

Signed-off-by: Neelakanta Venkatesh Petla <neelakanta.venkatesh.petla@intel.com>

venkateshpnv committed:
Merge branch 'main' of https://github.com/venkateshpnv/k8s-infra-offload-new into mgr_policy

sharanakumarpatil commented on pkg/inframanager/p4/policy_es2k.go:
```diff
@@ -0,0 +1,710 @@
+	return false
+}
+
+func IsSame(slice1 []uint16, slice2 []uint16) bool {
```
Is it better to move this func to utils? because using this in multiple files, reduces code duplication.

venkateshpnv replied:
Yes. All util functions have been moved to utils package.

sharanakumarpatil commented on pkg/inframanager/p4/policy_es2k.go:
```diff
@@ -0,0 +1,710 @@
+			tableName = "k8s_dp_control.acl_pod_ip_proto_table_egress"
+			if protocol != 0 {
+				entryAdd = p4RtC.NewTableEntry(
+					"k8s_dp_control.acl_pod_ip_proto_table_egress",
```
Is it better to use the tableName variable instead of the whole string?
If yes, then applicable to all NewTableEntry.

venkateshpnv replied:
Agree. For the time being I am making this change. But I have observed repetitive code at various places. I will ask Vishala to take a look.

sharanakumarpatil commented on pkg/inframanager/p4/policy_es2k.go:
```diff
@@ -0,0 +1,710 @@
+		}
+
+	case Delete:
+		tableName = "k8s_dp_control.acl_ipset_match_table_egress"
```
Duplicate line? tableName var getting initialized inside if/else cases.

venkateshpnv replied:
As mentioned about will ask Vishala to make these changes while looking at the repetitive code.

sharanakumarpatil commented on pkg/inframanager/p4/policy_es2k.go:
```diff
@@ -0,0 +1,710 @@
+		}
+
+	case Delete:
+		if protocol == 6 {
```
This can be combined with `if protocol == PROTO_TCP` using OR operation?

sharanakumarpatil commented on pkg/inframanager/p4/policy_es2k.go:
```diff
@@ -0,0 +1,710 @@
+				nil,
+			)
+		}
+		if protocol == 17 {
```
This can be combined with `if protocol == PROTO_UDP` ?

venkateshpnv replied:
Yes. Addressed the change.

dhruvpt commented:
Have we tested this change, and the existing features, to be working on both P4-DPDK and IPU platform?

venkateshpnv committed:
Rename ipsetid to rulegroup

Signed-off-by: Neelakanta Venkatesh Petla <neelakanta.venkatesh.petla@intel.com>

vishala-r commented on inframanager/api_handler/api_handler.go:
```diff
@@ -706,15 +720,274 @@ func (s *ApiServer) NatTranslationDelete(ctx context.Context, in *proto.NatTrans
+
+	policy = entry.(store.Policy)
+
+	err := p4.PolicyTableEntries(ctx, server.p4RtC, p4.PolicyDel, policy)
```
when you delete policy, ReleasePolicyIpsetIDX() should be taken care of properly. all acquired ipsetIdx should be returned back to the stack

venkateshpnv replied:
Will do. Thanks for pointing out.

vishala-r commented on inframanager/api_handler/api_handler.go:
```diff
@@ -706,15 +720,274 @@ func (s *ApiServer) NatTranslationDelete(ctx context.Context, in *proto.NatTrans
+	}
+	logger.Infof("Successfully added/updated policy %v to the pipeline", policy)
+
+	if ok := policy.WriteToStore(); !ok {
```
if its policy update, then better to call updateToStore()

venkateshpnv replied:
Done.

vishala-r commented on inframanager/api_handler/api_handler.go:
```diff
@@ -760,15 +1033,142 @@ func (s *ApiServer) RemoveHostEndpoint(ctx context.Context, in *proto.HostEndpoi
+		return out, err
+	}
+
+	if ok := workerEp.WriteToStore(); !ok {
```
call UpdateToStore() if its update call

venkateshpnv replied:
Ok. Will do. Thanks

vishala-r commented on pkg/inframanager/p4/policy_es2k.go:
```diff
@@ -0,0 +1,710 @@
+// Copyright (c) 2022 Intel Corporation.  All Rights Reserved.
```
this policy_es2k.go is not coded as per es2k p4. lut tables are not configured

venkateshpnv replied:
Currently this is just a dummy file to fix runner issues on git.

vishala-r commented on pkg/inframanager/store/storedefinition.go:
```diff
@@ -97,15 +103,17 @@ type IpSet struct {
 
 type PolicyWorkerEndPoint struct {
 	WorkerEp          string
```
whats the difference between workerEp and workerIp? either one of them should be removed right?

venkateshpnv replied:
workerEp is the name of the endpoint, where as the workerIp is the IP address of the endpoint. I have added the ip address to make it easy while programming the control plane.

vishala-r commented on pkg/inframanager/store/storedefinition.go:
```diff
@@ -141,8 +150,30 @@ func NewService() {
+			IpSetMap:      make(map[string]IpSet),
+			WorkerEpMap:   make(map[string]PolicyWorkerEndPoint),
+			PolicyLock:    &sync.Mutex{},
+			IpsetidxStack: utils.NewIpsetidxStack()}
```
probably this should also be renamed to rulesetid map or something

vishala-r commented on pkg/inframanager/store/storepolicy.go:
```diff
@@ -24,6 +24,14 @@ var (
 	WorkerepFile = StorePath + "workerep_db.json"
 )
 
+func GetNewPolicyIpsetIDX() int {
```
GetNewPolicyRuleSetID

venkateshpnv replied:
Done.

vishala-r commented on pkg/utils/utils.go:
```diff
@@ -69,6 +70,55 @@ var (
+		return false
+	}
+}
+
```
we have to add one more function to check if stack is full. its size should not go beyond 256. to avoid accidental pushes, we got to have that function.

venkateshpnv replied:
Added. Done.

vishala-r commented on pkg/utils/utils.go:
```diff
@@ -69,6 +70,55 @@ var (
+}
+
+func (st *IpsetidxStack) Push(value int) {
+	st.top++
```
and check here if stack is full...

venkateshpnv replied:
Added. Done.

venkateshpnv committed:
Policy code changes and cleanup

Signed-off-by: Neelakanta Venkatesh Petla <neelakanta.venkatesh.petla@intel.com>

venkateshpnv committed:
Merge branch 'main' of https://github.com/venkateshpnv/k8s-infra-offload-new into mgr_policy

venkateshpnv committed:
Updated temporary policy_es2k.go file

Signed-off-by: Neelakanta Venkatesh Petla <neelakanta.venkatesh.petla@intel.com>

venkateshpnv committed:
Addressed review comments and other fixes
- Rename ipsetidx to rulegroup
- Handle updatepolicy requests with stale policies
- Update stack implementation

Signed-off-by: Neelakanta Venkatesh Petla <neelakanta.venkatesh.petla@intel.com>

venkateshpnv committed:
Updated dummy policy_es2k.go file.

Signed-off-by: Neelakanta Venkatesh Petla <neelakanta.venkatesh.petla@intel.com>

venkateshpnv committed:
Optimize programming entries in the policy tables.

Signed-off-by: Neelakanta Venkatesh Petla <neelakanta.venkatesh.petla@intel.com>
